00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,483 @@
+
+There are alternative MSM algorithms that have better performance on native hardware, but those require either a higher number of memory accesses (which become additional lookups, with a corresponding additional expense), or leverage parallelism that isn't available to us in a zk context.
+
+### What is the impact of not doing this?
+
+It is not practical to build a Mina -> EVM state bridge without this or an similar alternative, given the practicalities of creating a proof using a more naive method.
+
+### Considerations regarding foreign field algorithm
+
+The RFC suggests going forward with [affine](https://www.hyperelliptic.org/EFD/g1p/auto-shortw.html) coordinates. However, [Jacobian](https://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#doubling-dbl-2007-bl) and [Projective](https://www.hyperelliptic.org/EFD/g1p/auto-shortw-projective.html) representations are also widely used in cryptography. Again, in practice affine coordinates will be most likely the most efficient for our scenario, and also easy to implement, so the suggested path is to start by implementing those, and revisiting the technique much later in the project timeline.
+
+
+## Prior art
+
+Regarding FF arithmetics and elliptic curve libraries, we might find inspiration looking at the following existing implementations (however, they are optimised for different scenarios, and generally quite complicated):
+- https://github.com/privacy-scaling-explorations/halo2wrong
+- https://github.com/DelphinusLab/halo2ecc-s
+- https://hackmd.io/@arielg/B13JoihA8
+
+
+To the best of our knowledge, our main task -- implementing a large MSM verification inside a SNARK -- has not been realised yet by anyone. Even though MSM algorithm exist in the aforementioned libraries, they are usually prohibitively expensive for our MSM size.
+
+<!--Discuss prior art, both the good and the bad, in relation to this proposal.
+
+Prior art is any evidence that your feature (invention, change, proposal) is already known.
+
+Think about the lessons from other blockchain projects or similar updates and provide readers of your RFC with a fuller picture. If there is no prior art, that is fine. Your ideas are interesting whether they are new or adapted from another source.-->
+
+## Unresolved questions
+
+1. Engineering problem: how exactly does the accumulation trick with the additive lookup tables work?
+   - If we go with the circuit layout that fits $2^{15}$ additions into $2^{15}$ rows having exactly one addition per row, we need to make sure the total joint accumulator (between folding iterations) is being updated properly. This requires altering the folding protocol checks.
+2. When splitting MSM into chunks --- is it not problematic that we have to enforce the right order of the chunks? The soundness of the approach relies heavily on the impossibility to swap these chunks around.
```

Comment:
The `buckets_input -> buckets_output` form of the proofs means that there's a canonical order, which is the only one that will work. Suggest deleting, and potentially inlining this note somewhere if important.

Reply:
OK that makes sense, but in the final folded instance, how do we know that we performed exactly `l = 17` buckets iterations, and not less? Unlike in zkVM, we don't have any code pointers...

Reply:
We're not folding at this stage.